File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	static int[] parent; 
	public static void main(String[] args) throws IOException {
		PriorityQueue<Edge> pq = new PriorityQueue<>();
		st = new StringTokenizer(br.readLine());
		int n = Integer.parseInt(st.nextToken());
		int m = Integer.parseInt(st.nextToken());
		for (int i=0; i<m; i++) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			int w = Integer.parseInt(st.nextToken());
			pq.offer(new Edge(a, b, w));
		}
		
		parent = new int[n+1];
		for (int i=1; i<=n; i++) {
			parent[i]=i;
		}
		
		int cnt =0 ;
		int result = 0;
		while (cnt < n-1) {
			Edge now = pq.poll();
			if (union(now.a,now.b)) {
				cnt++;
				result += now.w;
			}
		}
		System.out.println(result);
		
	}
	
	public static int find(int v) {
		if (v == parent[v])
			return v;
		return find(parent[v]);
	}
	public static boolean union(int a, int b) {
		a = find(a);
		b = find(b);
		if (a==b)
			return false;
		else if (a<b)
			parent[b] = a;
		else
			parent[a] = b;
		return true;
	}
	
	public static class Edge implements Comparable<Edge>{
		int a;
		int b;
		int w;
		public Edge(int a, int b, int w) {
			this.a = a;
			this.b = b;
			this.w = w;
		}
		@Override
		public int compareTo(Edge o) {
			return this.w-o.w;
		}
	}
}

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
import sys, heapq
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

v,e = map(int, input().split())
pq = [] 
parent = list(range(v+1))
for i in range(e):
    a,b,c = map(int, input().split())
    heapq.heappush(pq,(c,a,b))

def find(a):
    if a==parent[a]:
        return a
    else:
        parent[a] = find(parent[a])
        return parent[a]
    
def union(a, b):
    a = find(a)
    b = find(b)
    if a<b:
        parent[b] = a
    else:
        parent[a] = b

cnt = 0
result = 0
while cnt < v-1 and pq:
    w,s,e = heapq.heappop(pq)
    if find(s)!= find(e):
        union(s,e)
        cnt+=1
        result += w
print(result)